optimism/src/keccak/mod.rs

```diff
@@ -70,6 +79,77 @@ pub enum Constraint {
     IotaStateG(usize),
 }
 
+#[allow(dead_code)]
+/// The Keccak circuit
+pub(crate) type KeccakCircuit<F> = Circuit<ZKVM_KECCAK_COLS, Steps, F>;
+
+#[allow(dead_code)]
+impl<F: Field> KeccakCircuit<F> {
+    /// Create a new Keccak circuit
+    pub fn new(domain_size: usize) -> Self {
+        let mut circuit = Self {
+            witness: HashMap::new(),
+            constraints: Default::default(),
+            lookups: Default::default(),
+        };
+
+        for step in [
+            Round(0),
+            Sponge(Absorb(First)),
+            Sponge(Absorb(Middle)),
+            Sponge(Absorb(Last)),
+            Sponge(Absorb(Only)),
+            Sponge(Squeeze),
+        ] {
+            circuit.witness.insert(
+                step,
+                Witness {
+                    cols: Box::new(std::array::from_fn(|_| Vec::with_capacity(domain_size))),
+                },
+            );
+            circuit
+                .constraints
+                .insert(step, KeccakEnv::constraints_of(step));
+            circuit.lookups.insert(step, KeccakEnv::lookups_of(step));
+        }
+        circuit
+    }
+
+    /// Add a witness row to the circuit
+    pub(crate) fn push_row(&mut self, step: Steps, row: &[F; ZKVM_KECCAK_COLS]) {
+        // Make sure we are using the same round number to refer to round steps
+        let mut step = step;
+        if let Round(_) = step {
+            step = Round(0);
+        }
+        self.witness.entry(step).and_modify(|wit| {
+            for (i, value) in row.iter().enumerate() {
+                if wit.cols[i].len() < wit.cols[i].capacity() {
+                    wit.cols[i].push(*value);
+                }
+            }
+        });
+    }
+
+    /// Pads the rows of the witnesses until reaching the domain size
+    pub(crate) fn pad_rows(&mut self) {
```

Comment:
This is meant to pad the circuit with zeros until it reaches the domain size. I believe this is needed for folding, but probably not for non-folded circuits.